### PR TITLE
Support for dspy programs

### LINKED ISSUE #3: Add support for models with arbitrary dspy programs
Use `benchmark.py` to test various model + DSPy program combiantions

### PR DESCRIPTION
**Pull request by dkshjn**
closes #3

### CODE REVIEW

**Comment by sutyum** [comment]
A script (cli) which uses `benchmark.py` as a import (not as a cli).

Inside the repo with a dspy program (such as medprompt):
```bash
pip install git+https://github.com/Technoculture/med-llm-autoeval
```

```py
# benchmark_medprompt.py

import argparse
from benchmark import test
from medprompt import MedpromptModule

# class MedpromptModule(dspy.Module):
#   def __init__(self):
#     ...
# 
#  def forward(self, ...):
#     ...

if __name__ == "__main__":
  ... some argparse code ...
  
  results = test(
    dspy_module=MedpromptModule
    benchmark="openllm"
  )

  print(results)
```

> Both a prompt testing script (like medprompt.py, etc) and the benchmark.py can be used as cli scripts due to the code in their `if __name__ == ...` blocks as well as imported files.

**Comment by sutyum** [comment]
- [x] poetry to setup a project
- [x] benchmark.py - the main module in this library
- [ ] evauluate.py - python function to conduct an evaluation run
- [ ] scripts/example.py <-- how to use this library to write a script for testing a prompt module
- [x] Use ruff to format, lint and type check your code

**Review by sutyum — APPROVED** [review]

**Comment by sutyum** [comment]
Share a loom showing the script working

**Comment by dkshjn** [comment]
The following video explains the process.
https://www.loom.com/share/6c7e4eed82764d31b4bf4a6a859ac295?sid=160b0889-417f-4604-a758-5488df2b10e1

File: benchmark_medprompt.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import argparse
+from benchmark import test
+from medprompt import MedpromptModule
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
```
Comment (sutyum) [inline benchmark_medprompt.py:49]: none of these getting used?

File: medprompt.py
Comment (sutyum) [inline medprompt.py:1]: why is this file needed in this repo?

File: benchmark.py
Comment (sutyum) [inline benchmark.py:1]: Add a pydantic class to model TestConfig, also add an str enum for all the options when choosing the tests (one or many)

File: benchmark_medprompt.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    parser.add_argument("--model", type= str, default="gpt-3.5-turbo", help="Model to be used.")
+    parser.add_argument("--api_key", type=str, help="YOUR_API_KEY")
+    parser.add_argument("--benchmark", type=str, help = "Choose one of the following benchmark: [medmcqa, medicationqa, mmlu_medical, mmlu_general, arc, hellaswag, winogrande, blurb, truthfulqa, gsm8k].", default="arc")
+    parser.add_argument("--shots", type=int, help = "Number of few shots.", default=5)
+    parser.add_argument("--dspy_module", type = str, help = "Name of dspy module.", default="medprompt")
+    args = parser.parse_args()
```
Comment (sutyum) [inline benchmark_medprompt.py]: use the TestConfig pydantic parser here here:
```
args = TestConfig(**parser.parse_args())

results = test(args)
```

Comment (sutyum) [reply benchmark_medprompt.py]: Can also use this library for a seamless experience:
https://pypi.org/project/pydantic-argparse/

File: benchmark.py
Comment (sutyum) [inline benchmark.py:1]: use ruff for linting and formatter

**Review by sutyum — CHANGES_REQUESTED** [review]

File: benchmark_medprompt.py (hunk 1 above)
Comment (dkshjn) [reply benchmark_medprompt.py:49]: Instead of typing the parameters in the script itself you can run the script and  specify the parameters using the cli command also, for that reason this part is used. I chose to avoid that part in the video. If you want I can add the clip where I run the script using this method.

File: medprompt.py
Comment (dkshjn) [reply medprompt.py:1]: The medprompt module is described in this file. So we are importing the dspy program from here when running the script.

File: benchmark_medprompt.py (hunk 1 above)
Comment (sutyum) [reply benchmark_medprompt.py:49]: this script sample doesnt actually use the values passed to the cli currently

File: benchmark_medprompt.py (hunk 2 above)
Comment (dkshjn) [reply benchmark_medprompt.py]: Sure I will add this as well.

File: benchmark_medprompt.py (hunk 1 above)
Comment (dkshjn) [reply benchmark_medprompt.py:49]: Yes for the demo I entered the values that way, I will change that part.

**Review by sutyum — APPROVED** [review]

File: benchmark_medprompt.py (hunk 1 above)
Comment (0922640331) [reply benchmark_medprompt.py:49]: > Yes for the demo I entered the values that way, I will change that part.
